app/Http/Controllers/Form/WidgetSettingsController.php

```diff
@@ -0,0 +1,61 @@
+<?php
+/**
+ * WidgetSettingsController.php
+ *
+ * -Description-
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * @package    LibreNMS
+ * @link       http://librenms.org
+ * @copyright  2018 Tony Murray
+ * @author     Tony Murray <murraytony@gmail.com>
+ */
+
+namespace App\Http\Controllers\Form;
+
+use App\Http\Controllers\Controller;
+use App\Models\UserWidget;
+use Illuminate\Http\Request;
+
+class WidgetSettingsController extends Controller
+{
+    public function update(Request $request, $widget_settings)
+    {
+        $this->validate($request, ['settings' => 'array']);
```

Comment:
Doesn't validate the input data.

1. Create a top interfaces widget
2. Remove `type="number"` for `Last polled (minutes):`
3. enter "a" and save.

```
[2018-12-07 12:45:47] production.ERROR: ErrorException: A non-numeric value encountered in /opt/librenms/vendor/nesbot/carbon/src/Car
bon/Carbon.php:3639
Stack trace:
#0 /opt/librenms/vendor/nesbot/carbon/src/Carbon/Carbon.php(3639): Illuminate\Foundation\Bootstrap\HandleExceptions->handleError(2, '
A non-numeric v...', '/opt/librenms/v...', 3639, Array)
#1 /opt/librenms/app/Http/Controllers/Widgets/TopInterfacesController.php(55): Carbon\Carbon->subMinutes('a')
#2 /opt/librenms/app/Http/Controllers/Widgets/WidgetController.php(66): App\Http\Controllers\Widgets\TopInterfacesController->getView
(Object(Illuminate\Http\Request))
#3 [internal function]: App\Http\Controllers\Widgets\WidgetController->__invoke(Object(Illuminate\Http\Request))
#4 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Controller.php(55): call_user_func_array(Array, Array)
#5 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(44): Illuminate\Routing\Controller->callAct
ion('__invoke', Array)
#6 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Route.php(203): Illuminate\Routing\ControllerDispatcher->dispatch(Ob
ject(Illuminate\Routing\Route), Object(App\Http\Controllers\Widgets\TopInterfacesController), '__invoke')
#7 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Route.php(160): Illuminate\Routing\Route->runController()
#8 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(572): Illuminate\Routing\Route->run()
#9 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(30): Illuminate\Routing\Router->Illuminate\Routing\{clo
sure}(Object(Illuminate\Http\Request))
#10 /opt/librenms/app/Http/Middleware/VerifyTwoFactor.php(38): Illuminate\Routing\Pipeline->Illuminate\Routing\{closure}(Object(Illum
inate\Http\Request))
#11 /opt/librenms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(148): App\Http\Middleware\VerifyTwoFactor->handle(Obj
ect(Illuminate\Http\Request), Object(Closure))
#12 /opt/librenms/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Pipeline\Pipeline->Illuminate\Pipeline
\{closure}(Object(Illuminate\Http\Request))
...
```

One idea would be to add a `rules()` function to each widget controller and extend the validating with it,
for example `TopInterfacesController`:
```php
public function rules()
{
    return [
        'settings.interface_count' => 'required|integer|min:1',
        'settings.time_interval' => 'required|integer|min:1',
        'settings.interface_filter' => 'nullable'
    ];
}
```

Reply:
I would store the rules in the corresponding controller,
for example TopInterfacesController:
```php
public function rules()
{
    return [
        'settings.interface_count' => 'required|integer|min:1',
        'settings.time_interval' => 'required|integer|min:1',
        'settings.interface_filter' => 'nullable'
    ];
}
```
Then we need to figure out which controller to call from `WidgetSettingsController`, which is a bit trickier.
One way would be to just _assume_ it's called "CamelCaseController" based on the `widget`-field in the `Widget`-model. Since you talk about removing the whole Widget-table, maybe its a bad idea, and a hack, either way.
Removing the WidgetSettingsController seems like a way forward in that case